Bluetooth user authentication MVP

DimaWittmann opened the pull request:
## Overview

The purpose of this PR is to implement mean of authorization via ble. 
The spec can be found here: https://github.com/iotile/ArchWiki/wiki/Security-User-Authentication

As a first step ble internal encryption mechanism is used.
User and the device make a handshake as described in spec and as a result both has a session key.
The session key is used as out of band key to set up an encrypted connection.

## Major Changes

- Add core authentication logic, authentication is conducted after the characteristics were probed as additional stage of connection
- Add auth provider to prompt user password
- Implement handshake routines

### Additional Notes
3 new characteristics were added, two unidirectional characteristics that convey client/server hello and verify and one to signal if the device require authorization

How does it look from user perspective:
$ iotile hw --port=bled112
(HardwareManager) connect 0
(HardwareManager) get 8
(NRF52832Controller) test_interface
(ControllerTest) set_userpassword # set user password
Please, input user password for device  :
(ControllerTest) back
(NRF52832Controller) config_database
(ConfigDatabase) set_variable 'controller' 0x8002 uint8_t 1 # lock the device, aftert reset it will require authorization
(ConfigDatabase) back
(NRF52832Controller) reset
Resetting, takes at least 2 seconds
Please, input user password for device E6:0C:52:9A:D1:ED :
(NRF52832Controller) back
(HardwareManager) disconnect
(HardwareManager) connect 0
Please, input user password for device E6:0C:52:9A:D1:ED :
(HardwareManager)

mattrunchey commented on iotilecore/iotile/core/hw/auth/auth_provider.py:
```diff
@@ -46,147 +26,156 @@ def __init__(self, args=None):
-                key, however additional keys may be present depending on the encryption method
-                used.
+            bytearray: 16 bytes containing digest of key_purpose and reboot_counter
 
```
```suggestion
```

DimaWittmann replied:
Fixed

mattrunchey commented on iotilecore/iotile/core/hw/auth/auth_provider.py:
```diff
@@ -46,147 +26,156 @@ def __init__(self, args=None):
-                - sent_timestamp (int): The sent timestamp of the report
+    @classmethod
+    def DeriveRotatedKey(cls, reboot_key, current_timestamp, rotation_interval_power):
+        """ Derive an ephemeral key every 2^rotation_interval_power seconds,
```
```suggestion
        """Derive an ephemeral key every 2^rotation_interval_power seconds,
```

DimaWittmann replied:
Fixed.

mattrunchey commented on iotilecore/iotile/core/hw/auth/cli_auth_provider.py:
```diff
@@ -0,0 +1,35 @@
+class CliAuthProvider(RootKeyAuthProvider):
+    """Key provider implementation that prompt user for password
+
+        root key is derived using pbkdf2
```
```suggestion
       root key is derived using pbkdf2
```

DimaWittmann replied:
Got it.

mattrunchey commented on iotilecore/iotile/core/hw/auth/cli_auth_provider.py:
```diff
@@ -0,0 +1,35 @@
+        super().__init__(args)
+
+    @classmethod
+    def derive_key(cls, password):
```
no type hint or docstring

mattrunchey commented on iotilecore/iotile/core/hw/auth/cli_auth_provider.py:
```diff
@@ -0,0 +1,35 @@
+
+    @classmethod
+    def derive_key(cls, password):
+        return hashlib.pbkdf2_hmac('sha256', password.encode('utf-8'), b'salt', 100000)
```
Checking that this isn't a magic number -- why 100000? If there is a reason it's nice to at least comment it so we all remember.

mattrunchey commented on iotilecore/iotile/core/hw/auth/env_auth_provider.py:
```diff
@@ -1,226 +1,50 @@
-class EnvAuthProvider(AuthProvider):
-    """Authentication provider that can sign and verify using user keys
+class EnvAuthProvider(RootKeyAuthProvider):
+    """ Key provider implementation that search for keys in environment """
```
```suggestion
    """Key provider implementation that search for keys in environment """
```

Style consistency note (minor), most of the code has no space between the opening triple quote and the first word

mattrunchey replied:
Also want to clarify if the original docstring was meant to be deleted instead of updated

DimaWittmann replied:
The original was meant to be updated

mattrunchey commented on iotilecore/iotile/core/hw/auth/env_auth_provider.py:
```diff
@@ -1,226 +1,50 @@
 
-        var_name = "USER_KEY_{0:08X}".format(device_id)
+    def get_root_key(self, key_type, device_id):
+        """Attempt to get a user key from an environment variable"""
```
need to give more hints on what these variables are.

mattrunchey commented on iotilecore/iotile/core/hw/auth/auth_chain.py:
```diff
@@ -45,136 +44,33 @@ def _load_installed_providers(self):
-            NotFoundError: If the auth provider is not able to encrypt the data.
-        """
-
+    def get_root_key(self, key_type, device_id):
```
I got further down in the review and noticed that there are very similarly named functions that have docstrings in other classes (like base classes or other key modes I guess). 

I think that we still want to define the docstrings in these methods, because I can't think of a way to see the method we overwrote (and it's not an `abc`).

DimaWittmann replied:
Thank you for suggestions, you have already mentioned docstrings in your previous comments to other my MRs, I will strive to fix this style issues before passing it to review

mattrunchey commented on transport_plugins/bled112/iotile_transport_bled112/bled112.py:
```diff
@@ -800,7 +901,9 @@ def _on_disconnect(self, result):
         handle = context['handle']
 
-        callback(connection_id, self.id, success, "No reason given")
+        # TODO why self.id is used here, connection handle should be an argument?
```
I would prefer this TODO to be an open issue rather than buried in committed source code.

DimaWittmann replied:
I see your point, 
I added this TODO to draw attention during review, I am not sure if I should have done it in another way

DimaWittmann commented on iotilecore/RELEASE.md:
```diff
@@ -4,7 +4,19 @@ All major changes in each released version of `iotile-core` are listed here.
 
 ## HEAD
 
+## 5.0.9
```
I have two MRs pending, this is the second, the first one is https://github.com/iotile/coretools/pull/879, unfortunately I can't understand what is wrong is python37OSX version, maybe you could suggest how can I enabled debug for CI?

DimaWittmann commented on iotilecore/iotile/core/hw/auth/auth_provider.py:
```diff
@@ -1,33 +1,13 @@
-    Args:
-        args (dict): An optional dictionary of AuthProvider specific config information
-    """
+    """Base calls for all objects that provide a way to obtain a key"""
```
Yes, the functionality mentioned in comments was moved to other places.

DimaWittmann commented on iotilecore/iotile/core/hw/auth/auth_provider.py:
```diff
@@ -46,147 +26,156 @@ def __init__(self, args=None):
+            sent_timestamp (int): timestamp in report
+
+        Returns:
+            bytearray: 32 bytes  containing digest of id and timestamp
```
Fixed

DimaWittmann commented:
> A few missing files:
> 
>     * Need to update `version.py` in iotile-core
> 
>     * Need `RELEASE.md` and `version.py` from the BLED112 transport plugin
> 
> 
> The content/features in this PR are too complex for me to get a hold of and review in a reasonable amount of time, so the comments mostly revolve around style and some pointers to help clarify how the new features work (though I feel that we could always do more work there).
> 
> The link to the ArchWiki guide on User Authentication is nice -- I would suggest we consider moving the content of that wiki to a `.md` file somewhere in a coretools folder and update the root README with a "Security" section that simply links to that guide. It seems like the content is all generic enough to be open source, but we should verify.
> 
> A final "nice to have" would be some actual code examples of how you would write a coretools script to interact with encrypted systems, if possible.

@mattrunchey Thank you for the review, I added usage info to PR description. 
I added a hardware tests to this PR in pod_controller PR: https://github.com/iotile/pod_controller/pull/2
